### PR TITLE
Opt-in autocast for integers and floats

### LINKED ISSUE #9565: No auto-upcast for assignment of smaller ints to larger ones.
```crystal
def i(int : Int64)
end
v = 1
i(v) # Won't compile.
i(1) # Will compile.
```
The language should probably support auto-upcast of smaller ints to larger ones. The larger type is guaranteed to hold the value of the smaller one. This results in awkward casts and constants, for example the use of `1i64` to force a constant to a width.

The same problem exists for Float64 = Float32, and for Float = Int where the exponent is large enough to contain the Int.

I think I have also seen this called _type promotion_.

### PR DESCRIPTION
**Pull request by beta-ziliani**
Fixes #9565

Closes #9618

For the moment, it's guarded behind a compiler flag, in order to give us a time to test it. The following program compiles with `crystal build -Dnumber_autocast`:

```crystal
def foo(x : Int64)
  x
end

def bar
  1_i32
end

foo(bar)
```

### CODE REVIEW

File: spec/compiler/codegen/automatic_cast_spec.cr [hunk 1]
```diff
@@ -270,4 +270,15 @@ describe "Code gen: automatic cast" do
...
       )).to_i.should eq(255)
   end
+
+  it "casts integer variable to larger type (#9565)" do
+    run(%(
+      def foo(x : Int64)
+        x
+      end
+
+      x = 123
```
Comment (straight-shoota) [inline spec/compiler/codegen/automatic_cast_spec.cr:280]: Should we use an explicitly typed literal here? The test would no longer work if `Int64` became the default int type.

File: spec/compiler/semantic/automatic_cast_spec.cr [hunk 2]
```diff
@@ -573,9 +573,92 @@ describe "Semantic: automatic cast" do
       end
 
       Baz.new.as(Foo).foo(1)
-      )) { int64 }
+    )) { int64 }
   end
 
+  {% if flag?(:number_autocast) %}
```
Comment (straight-shoota) [inline spec/compiler/semantic/automatic_cast_spec.cr:579]: Why do we need this guard? The example gets evaluated with `number_autocast` flag, regardless of the spec compiler's flag.

File: src/compiler/crystal/semantic/ast.cr [hunk 3]
```diff
@@ -30,6 +30,58 @@ module Crystal
         false
       end
     end
+
+    # `number_autocast` defines if the feature in #9565 is present
```
Comment (straight-shoota) [inline src/compiler/crystal/semantic/ast.cr:34]:
```suggestion
    # `number_autocast` defines if automatic upcasting for number literals is enabled
```

File: src/compiler/crystal/semantic/ast.cr [hunk 4]
```diff
@@ -30,6 +30,58 @@ module Crystal
...
+        other_min <= self_min && self_max <= other_max
+      when {IntegerType, FloatType}
+        # Float32 mantissa has 23 bits,
+        # Float64 mantissa has 52 bits
+        case self_type.kind
+        when :i8, :u8, :i16, :u16
+          # Less than 23 bits, so convertable to Float32 and Float64 without precission loss
+          true
+        when :i32, :u32
+          # Less than 52 bits, so convertable to Float64 without precission loss
```
Comment (straight-shoota) [inline src/compiler/crystal/semantic/ast.cr:74]:
```suggestion
          # Less than 23 bits, so convertable to Float32 and Float64 without precision loss
          true
        when :i32, :u32
          # Less than 52 bits, so convertable to Float64 without precision loss
```

File: src/compiler/crystal/semantic/ast.cr (hunk 3 above)
Comment (beta-ziliani) [reply src/compiler/crystal/semantic/ast.cr:34]: Note that this is not just about number literals, but expressions.

File: spec/compiler/codegen/automatic_cast_spec.cr (hunk 1 above)
Comment (beta-ziliani) [reply spec/compiler/codegen/automatic_cast_spec.cr:280]: All and each of them. Also, this test should be tested with a flag.

File: src/compiler/crystal/semantic/ast.cr [hunk 5]
```diff
@@ -30,6 +30,58 @@ module Crystal
...
+          when FloatType
+            self_type.kind == :f32
+          end
+        else
+          false
+        end
+      end
+    end
+
+    def can_be_autocast_to?(other_type)
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/compiler/crystal/semantic/ast.cr:58]: Or `can_autocast_to?`

File: src/compiler/crystal/semantic/ast.cr (hunk 3 above)
Comment (straight-shoota) [reply src/compiler/crystal/semantic/ast.cr]: Right. "automatic upcasting for number values" then.

File: src/compiler/crystal/semantic/ast.cr (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sija) [reply src/compiler/crystal/semantic/ast.cr:58]: Yup, even better 👍

**Comment by HertzDevil** [comment]
Some possible additional specs:

```crystal
def foo(x : Int64 | UInt64)
  x
end

def bar
  1_i32
end

foo(bar) # should fail
```

```crystal
def foo(x : Int64 | String)
  x
end

def bar
  1_i32 || ""
end

foo(bar) # should be 1_i64 on codegen
```

**Comment by beta-ziliani** [comment]
@HertzDevil good examples, yet the second one fails (I don't honestly know why, since the mechanism is the same as for e.g. symbol autocast)

**Comment by asterite** [comment]
When I originally did this PR, I didn't add code to support that second example. So if you used that same code, it won't work. And, for now, I wouldn't worry about that case. Autocasting when unions are involved doesn't work at all, not even in the current autocasting:

```crystal
def foo(x : Int64 | String)
  puts x, typeof(x)
end

foo(1 || "a")
```

Gives:

```
In foo.cr:5:1

 5 | foo(1 || "a")
     ^--
Error: no overload matches 'foo' with type (Int32 | String)

Overloads are:
 - foo(x : Int64 | String)
Couldn't find overloads for these types:
 - foo(x : Int32)
```

**Review by straight-shoota — APPROVED** [review]

**Comment by Sija** [comment]
Would be nice to add these examples as `pending` anyhow.

**Comment by straight-shoota** [comment]
@Sija I don't think pending examples make sense here. That's just an independent, additional feature which isn't even completely well understood and finalized. We don't add pending specs for that. Instead, there should be an issue to track this feature request.

**Comment by asterite** [comment]
@beta-ziliani @straight-shoota Could we make this opt-out? That is, enable it by default with a way to stop this behavior if things go wrong.

Why?
- If we make it opt-in, nobody will use it. Nobody will know that this thing exists. Then you won't get any feedback about this, and you won't know whether to eventually make it the default.
- I'm almost sure this doesn't break existing code. And if it does, you can opt-out of it.

**Comment by beta-ziliani** [comment]
Sounds reasonable to me 👍

**Comment by straight-shoota** [comment]
Sure, makes sense to enable this for evaluation on nightly releases. We can reconsider before finalizing 1.3 whether we want to release it as opt-in. Until then we have at least a couple of weeks for testing.

**Comment by elorest** [comment]
I definitely prefer Opt-Out.

**Comment by straight-shoota** [comment]
How can you be using this feature for a year? It has just been implemented and merged a couple of weeks ago.

**Comment by elorest** [comment]
> How can you be using this feature for a year? It has just been implemented and merged a couple of weeks ago.

A couple years ago a feature was added which allowed for UInt8 to be passed in to params which were larger like UInt64. I have a couple lines of code which used to compile that don't now. It's nothing that can't be easily fixed though.

**Comment by Blacksmoke16** [comment]
@elorest I think that's moreso https://github.com/crystal-lang/crystal/pull/6074, since it's passing a number literal as an argument to the method, while this issue/PR is related to the case where you pass a variable/method return value as the argument.

**Comment by beta-ziliani** [comment]
> So maybe the opt out check was a bit too restrictive?
@Blacksmoke16 what do you mean here?

**Comment by Blacksmoke16** [comment]
@beta-ziliani Just that @elorest said:

> I have a couple lines of code which used to compile that don't now.

So was assuming the opt-in also made that existing feature opt-in.

**Pull request merged**